### ISSUE TITLE
CPU utilization and packet loss on high-throughput plans (500+Mbps) only.

### ISSUE THREAD

**Issue by rchac** [thebracket/cpumap-pping-hackjob#2]
I ran cpumap-pping in production for a bit and noticed it works very well for clients on speed plans of 250Mbps or less. Awesome work!

The only issue I was able to find is when a client is set to 500Mbps or greater. For these clients, when xdp_pping is run every 30 seconds or so, a few seconds into that operation the client on the high speed plan will experiences brief packet loss as the CPU core handling pping maxes out at 100%. The actual CPU core tied to the the traffic passing through maintains at a regular utilization. After xdp_pping finishes its process, CPU use returns to normal levels.

We do have a few clients on these 500Mbps plans (Ubiquiti WaveAP and fiber). In my case, TCP latency tracking isn't crucial - so I just switched to using xdp-cpumap-tc which resolved it.

![Screenshot from 2022-10-30 05-19-12](https://user-images.githubusercontent.com/22501920/198893766-2e349a9e-b7fc-420e-97c9-caad68d764ed.png)

![Screenshot from 2022-10-30 05-30-46](https://user-images.githubusercontent.com/22501920/198893769-9798ce1b-3ad1-480c-979b-6904ffbd8825.png)

The throughput was tested using Mikrotk btest on each end. Interestingly, if I set the transmission rate on the other end of the Mikrotik btest server to just below the enforced speed plan (instead of leaving that field blank), it does not trigger this issue! So some aspect of the HTB rate enforcement plus the ePPing part of cpumap-pping seem to interact here. Maybe when the ceiling is hit it breaks TCP and leads to xdp_pping having to do extra work?

This is not an urgent issue by any means (I'm back on xdp-cpumap-tc, all is well). Just reporting in case it's helpful to improve cpumap-pping long-term. =)

**Comment by dtaht** [comment]
sample every 10s? or 1s?

**Comment by dtaht** [comment]
It would kind of be my hope that @tohojo and @netoptimizer and dozens of others in the bpf world have a whole bunch of handy dandy perf an utperf flame graph one liners to see what's going on at a deeper level. please tell me all those influxdbs are running on a different core tho?

**Comment by rchac** [comment]
> please tell me all those influxdbs are running on a different core tho?

Just to see if that was involved, I went ahead and pinned InfluxDB to the last CPU core right now. Here's htop with InfluxDB pinned to the last core (11). First 3 cores are LibreQoS nodes. InfluxDB only briefly peaks at 4% when updating graphs then drops to ~1% on average. Surprisingly it doesn't influence CPU use much (at my ISP's small scale at least).

![image](https://user-images.githubusercontent.com/22501920/198925570-1a675825-8370-4560-ab2a-f7d7e8cef78a.png)

So I don't think InfluxDB is causing any major CPU concern.

**Comment by thebracket** [comment]
For larger installs, I do recommend putting Influx on a different host. It doesn't use a huge amount of CPU (it can be surprisingly RAM hungry sometimes) - but it can hit storage pretty hard. Ideally, you want it sitting on fast SSDs somewhere - and my experience is that storage is the first thing to wear out, so I prefer to have it somewhere less critical.

I'm actually not finding a huge amount in the way of eBPF profiling; lots of ways to *perform* profiling, but no common tool to manage it. It seems like a good idea to have a profiling tool (especially if it lets us catch the spikes), so I'll see what I can come up with.

**Comment by thebracket** [comment]
I've made some provisional changes in a branch, https://github.com/thebracket/cpumap-pping/tree/performance - they are looking great in limited testing, I'll try to test on a wider scale in the morning.

I've been referring to this as *sipping the firehose*. One major difference between the original *xdp-pping* and this project is that we're trying to gather data for a potentially huge amount of queues, rather than trying to accurately count whatever goes through. So it's ok to say "we've got some samples for this customer, let's worry more about the others". So there's some major changes in this branch:

1. Instead of overwriting a circular buffer per customer, we fill it until it hits the end of its assigned buffer space. At the *beginning* of the pping process, we check to see "does this customer have a full buffer yet?" - and if they do, it returns out - not doing any further work (not even decoding the packet). This will reset next time you run `xdp_pping`, so you still get a sample over time for each queue - but a really busy queue will a) no longer dominate the stats (making it less likely that we'll sample very quiet queues), and b) stop using any resources for monitoring once we've taken a sample.
2. The "time between samples" code has been changed to a simple sanity check. So it's biased towards getting the first data it can for a sampling period - which should be fine, since we're sampling in a periodic manner. This encourages the client's sample buffer to fill up rapidly, increasing the chances of (1) kicking in, and sampling no longer taking ANY resources for the remainder of the sampling period.
3. I've replaced the "packets" and "flows" buffers with LRU maps. LRU (Least Recently Used) maps never fill up - they replace the oldest member if you try to insert into a full buffer. That eliminates the need for the userspace `xdp_pping` to ever clean up these buffers - they have become self-cleaning, with old data naturally vanishing as it is replaced. It also makes the buffer size completely constant, allowing for a guaranteed allocation up-front - no allocations at runtime (and reduced memory fragmentation, as a side-effect).

With these changes, I'm seeing a slight performance increase in `iperf` testing between VMs (from about 4.5 Gb/s to 4.78 Gb/s in the last test between HyperV VMs). I'm also seeing lower CPU usage, because it stops trying to help after a flow has been sampled (hence "sipping the firehose" - we were trying to drink the whole thing, before).

I plan to try this on my network early tomorrow morning. If it doesn't explode, I'll let it run.

*Note that you have to reboot after compiling, the pinned maps have changed*

**Comment by rchac** [comment]
Awesome work! I'll try to test this out in production as soon as we hear back regarding your success with it.

**Comment by rchac** [comment]
Tested the performance branch. Working very well! Throughput for 500+ Mbps clients is also much more stable than before, presumably because there is no more CPU peaking. I think you resolved it. I'll leave it running. Fingers crossed. Awesome work!

TCP Latency graphing is coming together now.
![Screenshot from 2022-11-01 04-47-16](https://user-images.githubusercontent.com/22501920/199216895-9bb57116-228e-4af9-871d-29e4aa679ab6.png)

**Comment by thebracket** [comment]
Just got it applied here (running a little late due to a baby who wanted to play at 4am, but didn't want to get up at 6:30 to go to daycare..). So far, so good. A lot of our customers are students (it's that kind of town), so our heavy users aren't awake yet. So we're only pulling about 800 mbit/s so far.

CPU usage is great. The busiest core on the shaper box is at 22% (9 cores shaping on this box), quick speed tests on my workstation (which is hampered by a wifi AP that literally had the front fall off last night thanks to a construction worker!) is good, it feels nice and snappy.

Median latency reported on my network has actually gone up (still in the 40 ms range), while peak latency has gone down. That's expected, because of the previous version's oversampling of heavy users causing it to miss light use - which is often at the periphery where performance is hampered by going through a lot of backhaul hops. When I drill down, it's actually better for some of the nearby customers - and customers who weren't being sampled frequently now have data. So it's not really a like-for-like comparison.

I'd like to let it run, see how it does when the customers are awake!

I'll get this merged into the main branch today.

**Comment by thebracket** [comment]
Running this for 24 hours, and I found a subtle issue. Under very heavy packet-per-second load, I got some incorrect data: `xdp_pping` reported some TC handles that don't exist. I'm tracking this one down, it *feels* like a data race somewhere - but I'll have to do some instrumentation to be sure.

I'll wait on merging until the issue is resolved.

**Comment by thebracket** [comment]
Made it back from the doctor with the little one (she's ok, no sign of anything wrong - she's probably just teething). So I went ahead and pushed the latest set of changes. I *think* it'll fix the problem; if it doesn't, it's added the tools I need to figure out what's actually going wrong. I'll test first thing in the morning. It's a crazy-long commit message, so I'll just link to it: https://github.com/thebracket/cpumap-pping/commit/bbf264145f24247cd0dc587716f54ff59a470b0b

**Comment by rchac** [comment]
Thank goodness! Relieved to hear it.
And awesome. I'll test it out on my network tomorrow morning.

**Comment by thebracket** [comment]
Slow day, had to stay home with the little one (you can't go back to daycare until you've gone 24 hours without a high temperature. I swear her CPU is overheating from all the playing today).

Anyway, this turned out to be 99.9% fixed. The current performance tree very occasionally drops a packet rather than reporting bad data, now.

I also ran into a bug in BracketQoS that made this look a lot worse than it was. I'd forgotten to set a default. I highly recommend assigning a "default route" to the Trie, and putting it into an "unknown IPs" queue. That made things run a *lot* better:

In my case, the command is:

```
xdp_iphash_to_cpu_cmdline --add --ip ::/0 -cpu 8 --classid 2
```

I'm not 100% sure that Libre is handling default queues the same way, so you may want a different queue.

Do me a favor, if you wouldn't mind? Have a look at the output of `xdp_pping` and see if there are references to handles that don't exist. I'm still seeing occasional ones that I can't explain in BracketQoS - but not in my test instance of LibreQoS. So I'm hoping that's on my end, now.

**Comment by thebracket** [comment]
Actually, I think I just found the culprit hidden in some code I modified without sanity checking... I had a `memcpy` going 1 byte too far, overwriting some data. I have a test ready to go for the morning. (`memcpy` being one reason I wish this was in Rust!)
I haven't committed the change; want to test it locally first. It works on a tiny VM setup, I'd like to triple-check this one before anyone tries it in anger.

**Comment by rchac** [comment]
Sorry for delay - I just discovered my production server had Hyperthreading/SMT re-enabled somehow which caused kernel lockups. And ok sounds good. Once you give the clear to test i'll test it out.

**Comment by thebracket** [comment]
We're nearly there. :-) The last patch is working well on my setup. I saved 64-bits per TCP flow by not storing some redundant data (we already know the handle we are reporting on, why look it up?), fixed up a `memset` that was needlessly confusing and potentially writing just beyond the boundary of where it should have been writing. Performance is still good.

On my live BracketQoS instance, I'm still seeing it report an occasional TC handle that doesn't exist. I'm reasonably sure that these are IP addresses that haven't been added, but I'm still tracing it. I'd like to get that nailed down. It isn't doing any harm - I'm getting good results for the hosts that are setup correctly - but it is something that I'd like to tie down to avoid confusion.

**Comment by dtaht** [comment]
There are also totally bogus IP addresses to be had, bogons, rfc1918, bcp38 violators, and suchlike. send those somewhere?

**Comment by dtaht** [comment]
Having a default bogon, rfc1918, bad actors,  etc, "customer(s)" in the table might be an out. BOGONS, BADACTORS, RFC1918, people trying to reach the CGNAT address ranges 100.64.0.0/10  where you can't get there  - etc essentially tossed and logged...

**Comment by thebracket** [comment]
That's a good point (although they are only bogons if they are outside any
NAT, which complicates matters).

It actually opens up a whole discussion for 1.4: tracking internal ips we
don't recognize, so the isp can figure out what they forgot to map.

**Comment by thebracket** [comment]
I'm going to do a quick test in the morning of some cleaned-up code, and I *think* this is ready to merge. I've verified that the surprise queues are an artifact of how Bracket was handling unknown IP addresses. Ignoring those on graph input, performance is really good and I have graphs for customers. :-)

Edit: median monitored RTT when our upstream hit 1.85 Gbit/s last night (out of 2, so well into "you need to upgrade" land) was 42ms. If I exclude the one AP that currently hates the world, median was closer to 35ms. CPU usage never exceeded 45% on any of the 10 cores, despite a pretty imbalanced tree. (Not a very high-end shaper box, running in a VM - I'm pretty thrilled with that)

**Comment by rchac** [comment]
Awesome work! I've still had no issues even with the version from a few days back and I'm excited to test out this latest one.

Edit: Current graph setup

![Screenshot from 2022-11-04 15-36-10](https://user-images.githubusercontent.com/22501920/200078794-dd9c87c4-57a4-4573-8c6a-5ba25f8cf7f3.png)

**Comment by thebracket** [comment]
I really had to pound the crap out of it to find the issues that prompted the second half of this branch. That's one thing I run into from time to time; it's a heck of a lot easier to write code that works for a nice small test than it is when you hit it with real world pain!

I *really* like the bar distribution graph at the bottom-left. I always find I get a little lost in the massive number of overlayed lines, other than mouse over to see who the spike is. It's a nice visual way to get a feel for how it's doing - and the stacked bars give a much better feeling of "this is great". The part of me that enjoyed stats class couldn't help but notice that you have half a bell curve. :-)

**Comment by rchac** [comment]
Definitely, especially when it comes to something this complex.

And agreed! The bell curve shows up a little more clearly with more bins.
![Screenshot from 2022-11-04 17-02-31](https://user-images.githubusercontent.com/22501920/200086961-f08a7dad-e654-495d-8811-11cf2688d703.png)

I'll probably default it to displaying these histograms and the tables.

Super pleased with cpumap-pping, awesome work again. I can finally dial down on problem areas without sacrificing performance!

**Comment by thebracket** [comment]
The current version passed the torture test of a Friday night full of college students, so I've merged it. I'll close this issue. Thanks for the help!

**Comment by dtaht** [comment]
Does the color scheme on the histogram have any meaning?

**Comment by rchac** [comment]
The color scheme was due to grouping together circuits. Since it's not intuitive and just hammers the CPU to render that way we're back to a simpler histogram now:
![image](https://user-images.githubusercontent.com/22501920/200721296-0021cc6f-0f56-4f59-b2b7-c25a99bb9558.png)
